File: src/DatabaseManager/Table Edits/AddRowsEdit.cs
using System.Collections.Generic;
using System.Linq;

namespace DatabaseManager
{
    /// <summary>
    /// Represents a batch edit operation that adds multiple rows to a table.
    /// Each row is managed by an individual <see cref="AddRowEdit"/> instance.
    /// </summary>
    public class AddRowsEdit : TableEdit
    {

        private readonly AddRowEdit[] _addRowEdits;
        /// <summary>
        /// Gets the array of row edits representing the rows that were added.
        /// </summary>
        public AddRowEdit[] AddRowEdits
        {
            get
            {
                return _addRowEdits;
            }
        }
        /// <summary>
        /// Initializes a new instance of the <see cref="AddRowsEdit"/> class with a collection of row edits.
        /// </summary>
        /// <param name="rowEditsToAdd">An array of <see cref="AddRowsEdit"/> objects representing rows to add.</param>
        public AddRowsEdit(AddRowEdit[] rowEditsToAdd)
        {
            _addRowEdits = rowEditsToAdd;
        }
        /// <summary>
        /// Gets the original row data for each added row.
        /// </summary>
        /// <returns></returns>
        public List<object[]> GetOriginalRowEdits()
        {
            var edits = new List<object[]>(_addRowEdits.Length);
            foreach (AddRowEdit rowEdit in _addRowEdits)
                edits.Add(rowEdit.OriginalRowEdit);
            // 
            return edits;
        }

        /// <summary>
        /// Propagates a column addition to all added rows.
        /// </summary>
        /// <typeparam name="T">The type of the added column.</typeparam>
        /// <param name="indexOfColumn">The index where the column was added.</param>
        /// <param name="columnData">Optional data for the new column.</param>
        public override void ColumnAdded<T>(int indexOfColumn, T[] columnData = null)
        {
            foreach (AddRowEdit rowEditToAdd in _addRowEdits)
                rowEditToAdd.ColumnAdded(indexOfColumn, columnData);
        }

        /// <summary>
        /// Propagates a column deletion to all added rows.
        /// </summary>
        /// <param name="indexOfColumn">The index of the column to delete.</param>
        public override void ColumnDeleted(int indexOfColumn)
        {
            foreach (AddRowEdit rowEditToAdd in _addRowEdits)
                rowEditToAdd.ColumnDeleted(indexOfColumn);
        }

        /// <summary>
        /// Determines whether any added row contains the specified cell.
        /// </summary>
        /// <param name="indexOfColumn">The column index of the cell.</param>
        /// <param name="indexOfRow">The row index of the cell.</param>
        /// <param name="returnValue">The value of the cell if found.</param>
        /// <returns></returns>
        public override bool ContainsCell(int indexOfColumn, int indexOfRow, ref object returnValue)
        {
            foreach (AddRowEdit rowEditToAdd in _addRowEdits)
            {
                if (rowEditToAdd.ContainsCell(indexOfColumn, indexOfRow, ref returnValue))
                    return true;
            }
            return false;
        }

        /// <summary>
        /// Determines whether any added row contains the specified column.
        /// </summary>
        /// <param name="indexOfColumn">The column index to check.</param>
        /// <returns></returns>
        public override bool ContainsColumn(int indexOfColumn)
        {
            foreach (AddRowEdit rowEditToAdd in _addRowEdits)
            {
                if (rowEditToAdd.ContainsColumn(indexOfColumn))
                    return true;
            }
            return false;
        }

        /// <summary>
        /// Determines whether any added row is located at the specified index.
        /// </summary>
        /// <param name="indexOfRow">The row index to check.</param>
        /// <returns></returns>
        public override bool ContainsRow(int indexOfRow)
        {
            foreach (AddRowEdit rowEditToAdd in _addRowEdits)
            {
                if (rowEditToAdd.ContainsRow(indexOfRow))
                    return true;
            }
            return false;
        }

        /// <summary>
        /// Gets a list of all cell edits in the specified column across all added rows.
        /// </summary>
        /// <param name="indexOfColumn">The column index to retrieve edits from.</param>
        /// <returns></returns>
        public override List<CellEdit> GetEditedCellsInColumn(int indexOfColumn)
        {
            var result = new List<CellEdit>();
            foreach (AddRowEdit rowEditToAdd in _addRowEdits)
                result.AddRange(rowEditToAdd.GetEditedCellsInColumn(indexOfColumn));
            return result;
        }

        /// <summary>
        /// Gets a list of all cell edits in the specified row across all added rows.
        /// </summary>
        /// <param name="indexOfRow">The row index to retrieve edits from.</param>
        /// <returns></returns>
        public override List<CellEdit> GetEditedCellsInRow(int indexOfRow)
        {
            var result = new List<CellEdit>();
            foreach (AddRowEdit rowEditToAdd in _addRowEdits)
                result.AddRange(rowEditToAdd.GetEditedCellsInRow(indexOfRow));
            return result;
        }

        /// <summary>
        /// Updates the internal row index of all added rows when a new row is inserted in the table.
        /// </summary>
        /// <param name="indexOfRow">The index where the row was added.</param>
        /// <param name="rowData">Optional data for the new row.</param>
        public override void RowAdded(int indexOfRow, object[] rowData = null)
        {
            foreach (AddRowEdit rowEditToAdd in _addRowEdits)
                rowEditToAdd.RowAdded(indexOfRow, rowData);
        }

        /// <summary>
        /// Updates the internal row index of all added rows when a row is deleted from the table.
        /// </summary>
        /// <param name="indexOfRow">The index of the deleted row.</param>
        public override void RowDeleted(int indexOfRow)
        {
            foreach (AddRowEdit rowEditToAdd in _addRowEdits)
                rowEditToAdd.RowDeleted(indexOfRow);
        }
    }
}

File: src/DataBase_Reader_VB/Table Edits/AddRowsEdit.vb
Public Class AddRowsEdit
    Inherits TableEdit

    Private ReadOnly _addRowEdits() As AddRowEdit
    Public ReadOnly Property AddRowEdits() As AddRowEdit()
        Get
            Return _addRowEdits
        End Get
    End Property
    Public Sub New(rowEditsToAdd() As AddRowEdit)
        _addRowEdits = rowEditsToAdd
    End Sub
    Public Function GetOriginalRowEdits() As List(Of Object())
        Dim edits As New List(Of Object())(_addRowEdits.Count)
        For Each rowEdit As AddRowEdit In _addRowEdits
            edits.Add(rowEdit.OriginalRowEdit)
        Next
        '
        Return edits
    End Function
    Public Overrides Sub ColumnAdded(Of T)(indexOfColumn As Integer, Optional columnData() As T = Nothing)
        For Each rowEditToAdd As AddRowEdit In _addRowEdits
            rowEditToAdd.ColumnAdded(indexOfColumn, columnData)
        Next
    End Sub

    Public Overrides Sub ColumnDeleted(indexOfColumn As Integer)
        For Each rowEditToAdd As AddRowEdit In _addRowEdits
            rowEditToAdd.ColumnDeleted(indexOfColumn)
        Next
    End Sub

    Public Overrides Function ContainsCell(indexOfColumn As Integer, indexOfRow As Integer, ByRef returnValue As Object) As Boolean
        For Each rowEditToAdd As AddRowEdit In _addRowEdits
            If rowEditToAdd.ContainsCell(indexOfColumn, indexOfRow, returnValue) Then Return True
        Next
        Return False
    End Function

    Public Overrides Function ContainsColumn(indexOfColumn As Integer) As Boolean
        For Each rowEditToAdd As AddRowEdit In _addRowEdits
            If rowEditToAdd.ContainsColumn(indexOfColumn) Then Return True
        Next
        Return False
    End Function

    Public Overrides Function ContainsRow(indexOfRow As Integer) As Boolean
        For Each rowEditToAdd As AddRowEdit In _addRowEdits
            If rowEditToAdd.ContainsRow(indexOfRow) Then Return True
        Next
        Return False
    End Function

    Public Overrides Function GetEditedCellsInColumn(indexOfColumn As Integer) As List(Of CellEdit)
        Dim result As New List(Of CellEdit)
        For Each rowEditToAdd As AddRowEdit In _addRowEdits
            result.AddRange(rowEditToAdd.GetEditedCellsInColumn(indexOfColumn))
        Next
        Return result
    End Function

    Public Overrides Function GetEditedCellsInRow(indexOfRow As Integer) As List(Of CellEdit)
        Dim result As New List(Of CellEdit)
        For Each rowEditToAdd As AddRowEdit In _addRowEdits
            result.AddRange(rowEditToAdd.GetEditedCellsInRow(indexOfRow))
        Next
        Return result
    End Function

    Public Overrides Sub RowAdded(indexOfRow As Integer, Optional rowData() As Object = Nothing)
        For Each rowEditToAdd As AddRowEdit In _addRowEdits
            rowEditToAdd.RowAdded(indexOfRow, rowData)
        Next
    End Sub

    Public Overrides Sub RowDeleted(indexOfRow As Integer)
        For Each rowEditToAdd As AddRowEdit In _addRowEdits
            rowEditToAdd.RowDeleted(indexOfRow)
        Next
    End Sub
End Class
